pkg/promtail/targets/syslogtarget.go

```diff
@@ -0,0 +1,323 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/influxdata/go-syslog/v2"
+	"github.com/mwitkow/go-conntrack"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/relabel"
+
+	"github.com/grafana/loki/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/promtail/scrape"
+	"github.com/grafana/loki/pkg/promtail/targets/syslogparser"
+)
+
+var (
+	syslogEntries = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: "promtail",
+		Name:      "syslog_target_entries_total",
+		Help:      "Total number of successful entries sent to the syslog target",
+	})
+	syslogParsingErrors = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: "promtail",
+		Name:      "syslog_target_parsing_errors_total",
+		Help:      "Total number of parsing errors while receiving syslog messages",
+	})
+
+	defaultIdleTimeout = 120 * time.Second
+)
+
+// SyslogTarget listens to syslog messages.
+type SyslogTarget struct {
+	logger        log.Logger
+	handler       api.EntryHandler
+	config        *scrape.SyslogTargetConfig
+	relabelConfig []*relabel.Config
+
+	listener net.Listener
+	messages chan message
+
+	ctx             context.Context
+	ctxCancel       context.CancelFunc
+	openConnections *sync.WaitGroup
+}
+
+type message struct {
+	labels  model.LabelSet
+	message string
+}
+
+// NewSyslogTarget configures a new SyslogTarget.
+func NewSyslogTarget(
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	config *scrape.SyslogTargetConfig,
+) (*SyslogTarget, error) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	t := &SyslogTarget{
+		logger:        logger,
+		handler:       handler,
+		config:        config,
+		relabelConfig: relabel,
+
+		ctx:             ctx,
+		ctxCancel:       cancel,
+		openConnections: new(sync.WaitGroup),
+	}
+
+	t.messages = make(chan message)
+	go t.messageSender()
+
+	err := t.run()
+	return t, err
+}
+
+func (t *SyslogTarget) run() error {
+	l, err := net.Listen("tcp", t.config.ListenAddress)
+	l = conntrack.NewListener(l, conntrack.TrackWithName("syslog_target/"+t.config.ListenAddress))
+	if err != nil {
+		return fmt.Errorf("error setting up syslog target %w", err)
+	}
+	t.listener = l
+	level.Info(t.logger).Log("msg", "syslog listening on address", "address", t.ListenAddress().String())
+
+	t.openConnections.Add(1)
+	go t.acceptConnections()
+
+	return nil
+}
+
+func (t *SyslogTarget) acceptConnections() {
+	defer t.openConnections.Done()
+
+	l := log.With(t.logger, "address", t.listener.Addr().String())
+
+	backoff := util.NewBackoff(t.ctx, util.BackoffConfig{
+		MinBackoff: 5 * time.Millisecond,
+		MaxBackoff: 1 * time.Second,
+	})
+
+	for {
```

Comment:
```suggestion
	for backoff.Ongoing() {
```

Reply:
This leads to the ugly "use of closed network connection" error. I check for it inside the loop now: https://github.com/grafana/loki/pull/1275/files#diff-9e5d46359c7640bbb963370684c04406R119

Reply:
Ok, makes sense. In that case I'd suggest that we use `if t.ctx.Err() != nil {...}` instead of `if !backoff.Ongoing()` check., since we normally use `backoff.Ongoing()` with for loops.